### ISSUE TITLE
Handling warnings/errors for configuration and other non-rule issues

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#12650]
**The version of ESLint you are using.**

Latest

**The problem you want to solve.**

We are currently combining rule warnings/errors, parse errors, and some types of configuration errors in the lint messages object, separating as errors vs warnings but otherwise keeping them all together.

Since the warnings array in the lint result is counted and compared against the `--max-warnings` CLI option, this means that we sometimes find ourselves unable to add new configuration warnings (and other UX improvements to help our users configure and use ESLint correctly) due to not wanting to break users' builds on semver-minor releases.

I would like to allow us more flexibility on adding warnings for configuration or UX issues as semver-minor changes. Long-term, we need to be able to do this without having any impact on the exit code for the ESLint process.

This is currently blocking eslint/rfcs#34.

**Your take on the correct solution to problem.**

Two steps, one short-term and one long-term. (Note: The long-term plan is somewhat out of scope for this issue specifically, because it should actually be handled by an RFC. But I am hoping that describing the long-term approach at a high level will encourage the TSC to adopt the short-term approach for now.)

* **Short-term approach**: No technical change, just a process change which TSC could adopt via motion in TSC meeting.
  
  Basically, we would allow the TSC, on a strictly case-by-case basis, to allow specific proposals (which add new warnings) to be releasable in semver-minor releases. We would adopt specific guidelines and add them to the Maintainer Guide.
  
  As an initial proposal for those guidelines, I would suggest something like: (1) This shall not modify our current policy for warnings generated by ESLint rules; and (2) Anyone can add an issue to the TSC agenda and ask that the TSC consider accepting the proposal as semver-minor, despite adding more warnings from a non-rule source. But, we can discuss in more detail in the TSC meeting discussing this proposal. :smile:

* **Long-term approach**: Augment the lint result object to contain a top-level property for general lint run issues (including, but not limited to, configuration issues). This would contain warnings and errors arrays. Any errors in this object would result in exit code being nonzero (same as file-level errors, including rule errors). However, any warnings in this object would not result in exit code being nonzero. `--max-warnings` would not look at the warnings in this object-- it would only look at file-level warnings (ideally just rule warnings).
  
  Of course, this needs to be fleshed out in an RFC.

**Are you willing to submit a pull request to implement this change?**

Yes-- if the TSC agrees to the short-term approach, I would be willing to submit a PR to update the Migration Guide appropriately.

I could also work on the long-term RFC, but I am willing to let someone else take over that part if they are interested.

**Comment by platinumazure** [comment]
**TSC Summary**:

This is a follow-up from [a TSC discussion on 2019/12/05 about whether configuration-related warnings could ever be added in a semver-minor release](https://github.com/eslint/tsc-meetings/blob/master/notes/2019/2019-12-05.md#can-adding-warnings-without-adding-errors-ever-be-semver-minor).

I would like to allow us more flexibility on adding warnings for configuration or UX issues as semver-minor changes. Long-term, we need to be able to do this without having any impact on the exit code for the ESLint process.

More background can be found in the initial issue post. Also, this is currently blocking [eslint/rfcs#34](https://github.com/eslint/rfcs/pull/34).

Long-term, we should probably fix this by creating a new location in the lint result object for lint run-level issues (e.g., configuration issues). Short-term, I would like to ask the TSC to consider allowing exceptions to the semver rules, on a case-by-case basis, for specific proposals in which the value of adding warnings is high enough to potentially justify breaking the users' build on semver-minor releases (especially if ESLint is already currently not working for the use case in question and we want to improve the UX and help users fix their issue). The process would be to nominate the issue on the TSC Agenda and ask the TSC to decide if the proposal should be accepted while adding warnings on a semver-minor release.

This process change would be captured in the Maintainer Guide.

**TSC Question**:

Should we adopt a policy to allow, on a case-by-case basis, TSC discussion on issues that add warnings and whether they could be allowed in a semver-minor release?

**Comment by mysticatea** [comment]
I think, adding warnings in minor releases is fine if the warning reports existing *wrong* behaviors. It will improve UX with more understandable messages.

On the other hand, we should be careful if the warning reports existing *right* behaviors. Because it consequently means deprecation of existing features. Our policy allows deprecation of rules in minor releases, but about deprecation of core features is undefined. The change of runtime behaviors by deprecation will surprise users. I think it's better if adding the warnings which report existing right behaviors happens in major releases. For a prior art, [Node.js Deprecations](https://github.com/nodejs/node/blob/master/COLLABORATOR_GUIDE.md#deprecations).

**Comment by platinumazure** [comment]
> I think, adding warnings in minor releases is fine if the warning reports existing *wrong* behaviors. It will improve UX with more understandable messages.

@mysticatea I just want to clarify this by asking for a specific example:

For the purpose of RFC 34, where we want to add a warning for ESLint reporting something that it had previously been ignoring: Is this a wrong behavior? (ESLint is not doing what the user expects right now.)

Thanks for commenting!

**Comment by mysticatea** [comment]
Indeed, "the user expects" is ambiguous. In the RFC34 case, the kind of comments is not well-documented, so it's hard to say what is right.

But, I feel `// global ...` and `// eslint ...` are existing right behaviors. For example, the following comment is reasonable to explain the reason for disabling:

```js
    // eslint has false positive: https://github.com/eslint/eslint/issues/11899
    // eslint-disable-next-line require-atomic-updates
    obj.value = 0;
```

Also `globals` and `exported` may be at the beginning of line comments as valid sentences. (please mind there is a style to represent multiline comments by consecutive line comments and break lines by fixed width.)

Therefore, I think that we should deprecate "line comments that begin by the words 'eslint', 'global', 'globals', and 'exported'" and announce it.

**Comment by kaicataldo** [comment]
I still think it would helpful to differentiate between "warnings that are errors in rules" and "logged information to improve the UX of ESLint from ESLint itself". We could then update our semver policy to say that we reserve the right to add the latter type of warning in semver-minor releases.

**Comment by mysticatea** [comment]
> I still think it would helpful to differentiate between "warnings that are errors in rules" and "logged information to improve the UX of ESLint from ESLint itself". 

I agree. There is a precedent. Linter returns `usedDeprecatedRules` property along with the lint results and it contains the ID and substitute of used deprecated rules. The logged information may merge it.

That said, I think that it's surprising behavior if the example in https://github.com/eslint/eslint/issues/12650#issuecomment-563956133 raises a warning in a minor release. That looked completely legal.

**Comment by kaicataldo** [comment]
Reopening because we still need to discuss this. @eslint/eslint-tsc Please give this a look when you get a chance!

**Comment by sam3k** [comment]
A short status summary:

- This issue has been reopened for two years
- Does not seem to be part of the TSC agenda at the moment
- Has an `evaluating` label but doesn't look like that's the case?

**Comment by nzakas** [comment]
Thanks @sam3k! Yeah, I don't think we are going to move forward with any changes here.

It seems like we kind of solved this problem by using `process.emitWarning()` for non-lint errors and warnings:
https://github.com/eslint/eslint/blob/d94dc84ae76a36b4ee9268c40d8536d2f5b1c63c/lib/shared/deprecation-warnings.js
